url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/EvalTrace.java

```diff
@@ -0,0 +1,921 @@
+
+            public void receivedException(Eval2.EvalEvent aEvent)
+            {
+                synchronized (EvalTrace.this)
+                {
+                    halted = true;
+                    stack = aEvent.getStack();
+                    exception = aEvent.getException();
+                    excepted = true;
+                    result = null;
+                    hasResult = false;
+                    // Signal the exception event to any processes
+                    // that are waiting for something to happen.
+                    EvalTrace.this.notifyAll();
+                }
+            }
+        });
+
+        // Create a separate suspendable thread to evaluate our expresion in.
+        worker = new Thread()
+        {
+            @Override
+            public void run()
+            {
+                try
+                {
+                    eval.eval(aExpr);
+                }
+                catch (CommandException e)
+                {
+                    // The eval will have notified any listeners.
+                    // Our EvalTrace will have received the exception event and should have
+                    // taken appropriate action.
+                    // We can let this thread die.
+                }
+            }
+        };
+
+        // Start the eval thread.
+        worker.setName("eval/trace");
+        worker.setDaemon(true);
+        worker.start();
+    }
+
+    public synchronized boolean hasMoreSteps()
+    {
+        if(worker == null || excepted || !worker.isAlive() ) return false;
+
+        try
+        {
+            while (!halted) this.wait();
+            // This is a difficult piece of code.
+            // We halt if there is an exception, this is the easy part.
+            // We also halt if the stack has no more steps AND there is not previous stack.
+            // There can be multiple stacks if there is a nesting, i.e. a command invokes eval itself.
+            // In the nested case, the debugger can go on stepping, no problem.
+            return stack != null && !excepted && (stack.hasMoreSteps() || stack.getPrevStack() != null);
+        }
+        catch (InterruptedException e)
+        {
+            halted = true;
+            return false;
+        }
+    }
+
+    public synchronized void step()
+    {
+        if(worker == null || excepted || !worker.isAlive()) throw new IllegalStateException();
+
+        try
+        {
+            while (!halted) this.wait();
+            if(!excepted)
+            {
+                // Clear the breakpoint flag.
+                breakpoint = false;
+
+                // Wake up the eval to take a step for us.
+                this.notifyAll();
+                // We put ourselves in wait mode until the eval has taken a step.
+                // The eval has to get us out of the wait queue.
+                this.wait();
+                stepcount ++;
+            }
+        }
+        catch (InterruptedException e)
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)